### PR TITLE
address book

### PR DESCRIPTION
**Pull request by RustemYuzlibaev**

### CODE REVIEW

File: src/components/AddressBook/AddressBookDialog.vue [hunk 1]
```diff
@@ -0,0 +1,275 @@
+<template>
+  <dialog-base :title="'Your contacts'" :visible.sync="isVisible">
+    <div v-if="userHasContacts">
+      <search-input
+        autofocus
+        v-model="search"
+        :placeholder="'Search by address, name or identity'"
```
Comment (stefashkaa) [inline src/components/AddressBook/AddressBookDialog.vue:7]: i18n please

File: src/components/AddressBook/AddressBookDialog.vue [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+      </s-button>
+    </div>
+  </dialog-base>
+</template>
+
+<script lang="ts">
+import { api } from '@sora-substrate/util';
+import { Component, Mixins, Ref } from 'vue-property-decorator';
+
+import type { AppWallet } from '@/consts';
```
Comment (stefashkaa) [inline src/components/AddressBook/AddressBookDialog.vue:61]: please don't use relative imports here (with @)

File: src/components/AddressBook/AddressBookDialog.vue [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+    this.closePopover();
+    this.closeDialog();
+  }
+
+  async mounted(): Promise<void> {
+    await this.updateAddressBook();
+    await this.selectWallet(this.selectedExtension as AppWallet);
+    this.extensionAccounts = (await this.getFormattedExtensionList(this.polkadotJsAccounts)).filter(this.filterRecord);
+
+    this.$root.$on('updateAddressBook', () => {
```
Comment (stefashkaa) [inline src/components/AddressBook/AddressBookDialog.vue:213]: do you really need root events?

File: src/components/AddressBook/Options.vue [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+    <div slot="reference" @click="handleOpenOptions">
+      <s-icon class="address-book__option-icon" name="basic-more-vertical-24" />
+    </div>
+  </el-popover>
+</template>
+
+<script lang="ts">
+import { Component, Mixins, Prop } from 'vue-property-decorator';
+
+import type { AccountBook } from '@/types/common';
```
Comment (stefashkaa) [inline src/components/AddressBook/Options.vue:24]: the same comment regarding imports

File: src/components/AddressBook/SetContactDialog.vue [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+      </div>
+    </div>
+  </dialog-base>
+</template>
+
+<script lang="ts">
+import debounce from 'lodash/debounce';
+import { Component, Mixins, Prop, Ref, Watch } from 'vue-property-decorator';
+
+import { state, getter, mutation } from '@/store/decorators';
```
Comment (stefashkaa) [inline src/components/AddressBook/SetContactDialog.vue:41]: +

File: src/components/AddressBook/SetContactDialog.vue [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+  @state.account.polkadotJsAccounts polkadotJsAccounts!: Array<PolkadotJsAccount>;
+  @getter.account.account account!: PolkadotJsAccount;
+  @mutation.account.setAddressToBook setAddressToBook!: (record: AccountBook) => void;
+
+  @Prop({ default: '', type: String }) readonly prefilledAddress!: string;
+  @Prop({ default: false, type: Boolean }) readonly isEditMode!: boolean;
+
+  @Ref('address') private readonly addressInput!: HTMLInputElement;
+
+  @Watch('isVisible')
```
Comment (stefashkaa) [inline src/components/AddressBook/SetContactDialog.vue:68]: do you really need another watcher? you already have `input`/`change` events of s-input component

File: src/components/AddressBook/SetContactDialog.vue [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+      this.name = this.book[this.prefilledAddress];
+    }
+
+    await this.$nextTick();
+    this.addressInput.focus();
+  }
+
+  @Watch('address')
+  handleAddressInput(address: string): void {
+    if (!api.validateAddress(address)) {
```
Comment (stefashkaa) [inline src/components/AddressBook/SetContactDialog.vue:87]: [87-90] lines should also be included into debounce function cuz `api.validateAddress` is complex operation and it's better not to execute it per each keyboard interaction

File: src/components/AddressBook/SetContactDialog.vue [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+
+  address = '';
+  name = '';
+  onChainIdentity = this.t('addressBook.none');
+
+  defineIdentity = debounce(this.getIdentity, 500);
+
+  setContact(): void {
+    this.setAddressToBook({ address: formatSoraAddress(this.address), name: this.name });
+    this.$root.$emit('updateAddressBook');
```
Comment (stefashkaa) [inline src/components/AddressBook/SetContactDialog.vue:103]: do you really need root events here?

File: src/components/AddressBook/SetContactDialog.vue [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+        <s-button @click="setContact" type="primary" class="s-typography-button--large" :disabled="btnDisabled">
+          {{ btnText }}
+        </s-button>
+      </div>
+    </div>
+  </dialog-base>
+</template>
+
+<script lang="ts">
+import debounce from 'lodash/debounce';
```
Comment (stefashkaa) [inline src/components/AddressBook/SetContactDialog.vue:38]: `lodash/fp/debounce` -> to reduce number of exported files (check rollup config please)

File: src/components/AddressBook/SetContactDialog.vue (hunk 8 above)
Comment (RustemYuzlibaev) [reply src/components/AddressBook/SetContactDialog.vue:103]: it's useful as SetContactDialog.vue and AddressBookDialog.vue are separate components, once you set new contact in SetContactDialog.vue (also setting in localStorage), you have to update the current list in AddressBookDialog.vue  and resort it alphabetically.

File: src/components/AddressBook/Options.vue [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+  get computedClass(): string {
+    return this.withActiveOptions ? 'address-book__option' : 'address-book__option--not-active';
+  }
+
+  handleSelectAddress(address: string, name: string): void {
+    this.$emit('select-address', address, name);
+  }
+
+  handleOpenOptions(event: Event): void {
+    event.stopPropagation();
```
Comment (Nikita-Polyakov) [inline src/components/AddressBook/Options.vue:42]: I can't undestand the purpose of this method. Could it be removed?

**Review by Nikita-Polyakov — APPROVED** [review]

**Pull request merged**